### PR TITLE
ikfast: implement Translation*AxisAngle4D IK type

### LINKED ISSUE #548: IkParameterizationType translationyaxisangle4d is not implemented
### TranslationYaxis4D is not implemented in moveit_kinematics

With the help of @gavanderhoorn his answer on http://answers.ros.org/question/263925/generating-an-ikfast-solution-for-4-dof-arm/ I have made an IkFast plugin for my robot.
When dragging the handles in Moveit the robot goal pose doesn't move, and while trying to drag I get the message:
```
[ERROR] [1500016582.397196842]: IK for this IkParameterizationType not implemented yet.
```

I've looked this up and came to the following answer on ros.answers.org http://answers.ros.org/question/191884/ik-for-this-ikparameterizationtype-not-implemented-yet/

I would be greatly helped to know which actions I should take to (similarly as in the answer above) get these kinematics implemented.

### Environment
* ROS Distro: Kinetic
* OS Version: Ubuntu 16.04

### PR DESCRIPTION
**Pull request by mintar**
Added an implementation for TranslationXAxisAngle4D TranslationYAxisAngle4D and
TranslationZAxisAngle4D. @nxdefiant tested this with his robot arm for the
TranslationZAxisAngle4D IK type (see #548); the rest should be analogous.

Can be cherry-picked onto kinetic.

Fixes #548 .

- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Document API changes relevant to the user in the moveit/MIGRATION.md notes
- [ ] Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [x] Decide if this should be cherry-picked to other current ROS branches
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by nxdefiant** [comment]
At least I can not use Translation*AxisAngle*Norm4D..
Also I think "identify" is the wrong word. If both do share the same implementation why not handle the two in the same switch-case?

**Comment by rhaschke** [comment]
@nxdefiant, what kind of errors do you get when trying to use `Translation*AxisAngle*Norm4D`?

> If both do share the same implementation why not handle the two in the same switch-case?

I agree, that this is a reasonable approach if they share the same implementation. But right now, I cannot see any difference between the two, which means that they _are identical_. Is the _ik type_ used in `ComputeIk()` to compute different things? I haven't looked this up, I'm just asking.

**Comment by nxdefiant** [comment]
Well I don't say I know what I'm doing. But when creating the ik solver with OpenRave for [this urdf](https://github.com/ROBOTIS-GIT/open_manipulator/blob/master/open_manipulator_description/urdf/open_manipulator.urdf.xacro)

it works with TranslationZAxisAngle4D:

> openrave.py --database inversekinematics --robot=open_manipulator.xml --usecached --iktests=100 --iktype=TranslationZAxisAngle4D
(...)
> openravepy.databases.inversekinematics: testik, success rate: 1.000000, wrong solutions: 0.000000, no solutions: 0.000000, missing solution: 0.010000

but fails for TranslationZAxisAngleYNorm4D:

> openrave.py --database inversekinematics --robot=open_manipulator.xml --usecached --iktests=100 --iktype=TranslationZAxisAngleYNorm4D
> openravepy.databases.inversekinematics: testik, success rate: 0.120000, wrong solutions: 0.730000, no solutions: 0.150000, missing solution: 0.080000
Traceback (most recent call last):
  File "/usr/local/bin/openrave.py", line 126, in <module>
    database.run(args=args)
  File "/usr/local/lib/python2.7/dist-packages/openravepy/_openravepy_0_9/databases/inversekinematics.py", line 1123, in run
    InverseKinematicsModel.RunFromParser(*args,**kwargs)
  File "/usr/local/lib/python2.7/dist-packages/openravepy/_openravepy_0_9/databases/inversekinematics.py", line 1111, in RunFromParser
    raise InverseKinematicsError(u'wrong rate %f > 0!'%wrongrate)
openravepy._openravepy_0_9.databases.inversekinematics.InverseKinematicsError: Inverse Kinematics Error
wrong rate 0.730000 > 0!

This is OpenRave branch latest_stable (commit 9c79ea26) on an ubuntu 16.04 Docker Image just for this task.

**Comment by mintar** [comment]
@rhaschke : Thanks for the quick review!

To clarify your question: the `ComputeIk` function is part of the `<robotarm>_ikfast_solver.cpp` file that's generated by OpenRAVE (e.g.: `openrave.py [...] --iktype=TranslationZAxisAngle4D`). Depending on the `--iktype` given to OpenRAVE during generation time, a different implementation of `ComputeIk` will be generated. These implementations differ in the arguments that `ComputeIk` expects in the second argument (rotation), for example:

* `Transform6D` expects the second argument of ComputeIk to be an array of 9 values for the 3x3 rotation matrix.
* `TranslationZAxisAngle4D` and `TranslationZAxisAngleYNorm4D` both expect the second argument of ComputeIk to be a pointer to a single float, the pitch angle.

This difference is why we need the switch statement in the first place, but it also means that we can handle different IK types in the same switch case if they have the same interface.

TL;DR: Yes, `ComputeIk` differentiates between solver types (or rather: different solvers will be generated depending on which solver type the user selected during generation time).

**Comment by rhaschke** [comment]
Thanks for the clarification. I'm going to merge (and cherry-pick) this simple PR without a 2nd review.
